atc/metric/emitter/influxdb.go

```diff
@@ -60,34 +81,56 @@ func (emitter *InfluxDBEmitter) Emit(logger lager.Logger, event metric.Event) {
 		return
 	}
 
-	tags := map[string]string{
-		"host": event.Host,
-	}
-
-	for k, v := range event.Attributes {
-		tags[k] = v
+	for _, event := range events {
+		tags := map[string]string{
+			"host": event.Host,
+		}
+
+		for k, v := range event.Attributes {
+			tags[k] = v
+		}
+
+		point, err := influxclient.NewPoint(
+			event.Name,
+			tags,
+			map[string]interface{}{
+				"value": event.Value,
+				"state": string(event.State),
+			},
+			event.Time,
+		)
+		if err != nil {
+			logger.Error("failed-to-construct-point", err)
+			continue
+		}
+
+		bp.AddPoint(point)
 	}
 
-	point, err := influxclient.NewPoint(
-		event.Name,
-		tags,
-		map[string]interface{}{
-			"value": event.Value,
-			"state": string(event.State),
-		},
-		event.Time,
-	)
-	if err != nil {
-		logger.Error("failed-to-construct-point", err)
-		return
-	}
-
-	bp.AddPoint(point)
-
 	err = emitter.client.Write(bp)
 	if err != nil {
 		logger.Error("failed-to-send-points",
 			errors.Wrap(metric.ErrFailedToEmit, err.Error()))
 		return
 	}
+	logger.Info("influxdb-emitter-fork-influxdb-batch-emitted", lager.Data{
+		"size": len(events),
+	})
+}
+
+
+func (emitter *InfluxDBEmitter) Emit(logger lager.Logger, event metric.Event) {
+	batch = append(batch, event)
+	duration := time.Since(lastBatchTime)
+	if len(batch) > emitter.batchSize || duration > emitter.batchDuration {
+		logger.Debug("influxdb-pre-emit-batch", lager.Data{
+			"influxdb-batch-size": emitter.batchSize,
+			"current-batch-size": len(batch),
+			"influxdb-batch-duration": emitter.batchDuration,
+			"current-duration": duration,
+		})
+		go emitBatch(emitter, logger, batch)
```

Comment:
I'm not very sure about making each emission a goroutine 🤔 If I understood correctly, this makes the buffer "infinite" as there's never really going to exist something that throttles the writes. Is this statement right? Maybe I got something wrong 😬

Reply:
Making it a goroutine ensures that the construction of the batch points and writing to InfluxDB are immediately taken off the line of execution so that the emitter can continue to build up the next batch.

And this existing chunk of code in emit.go still ensures that events are passed on in a synchronous way to the InfluxDBEmitter after they have been read from the channel.

`func emitLoop() {
	for emission := range emissions {
		emission.logger.Debug("emit-event-loop", lager.Data{
			"event": emission.event,
		})
		emitter.Emit(emission.logger.Session("emit"), emission.event)
	}
}`

So there will only ever be one active batch in progress that will be submitted when either the size or duration limit is reached.

Does this answer your question? Or am I missing something?